enhancements/update/eus-upgrades-mvp.md

```diff
@@ -393,6 +390,31 @@ The idea is to find the best form of an argument why this enhancement should _no
 
 ## Alternatives
 
+### MCO Kubelet Skew Enforcement
+
+Instead of the API Server Operator enforcing kubelet skew compliance through
+the `Upgradeable` flag, the MCO could provide this functionality. Either of
+these two operators are the obvious choice for such a check since they are
+responsible for both halves of the kubelet-API Server interaction. It makes
+more sense for the leading component to implement the check, however, since
```

Comment:
I think this is a critical point for discussion. The two arguments in the other direction are locality to the resolving action and locality of skew policy knowledge.

locality to the resolving action is about where a user has to go to resolve an Upgradeable=False condition.  The natural assumption we've seen from users and support in the past is that the operator with the unexpected condition is where to take a resolving action.  Now in this particular case, the kube-apiserver will be the spot for a resolving action 0% of the time. The kube-apiserver will never be able to upgrade a kubelet to close the skew.  The MCO will sometimes (often?) be the spot to close skew.  Either by waiting for a rollout or by unpausing a machine-config pool.  It's not perfect (byo-host for instance), but it is significantly more than the 0% of the kube-apiserver.

locality of skew policy knowledge is about which component knows its skew policy.  It's worth remembering that even in a kube control plane, skew policy between kubelet, kube-controller-manager, kube-scheduler, SDN (and its various variants) are different.  Expanding to everything in our payload (and what they subsequently install) it gets even larger.  While it is possible for the kube-apiserver to be a rally point for every team enforcing their skew policy, the distributed nature of clusteroperator conditions means that knowledge can be codified close to the component with the skew policy knowledge.  In this case, the MCO is the closest component (since we lack a kubelet operator).

In the general case, the locality to the resolving action and the locality to the skew policy knowledge aligns extremely well.  In this specific case, they meet in the MCO.

I've taken the liberty of demonstrating in a PR https://github.com/openshift/machine-config-operator/pull/2552

Reply:
I buy the first argument, but I'm struggling with "locality of skew policy knowledge". Between the kubelet and the API Server, it would be the API Server that dictates the skew policy based on its adherence to a backward-compatible API. This boils down to the same argument I've made in the PR: the leading component determines the acceptable skew. Based on that understanding, I would think the API Server Operator would be the place to at least store the skew policy.

Reply:
> it would be the API Server that dictates the skew policy based on its adherence to a backward-compatible API

This is the point: the kube-apiserver dictates the API versions served only, the clients know which API versions they need and depend on. For that reason in upstream the API-Machinery team is **NOT** the one responsible for testing and fixing the skew of a release, but node team is. The API-Machinery team just follows the known, mechanical procedures of introducing new API versions, deprecating and removing old ones.

Reply:
Or in yet other words:
- locality to resolve skew problem: **UX problem**
- locality to enforce skew policy: **ppl problem**, who owns the skew policy (kube-apiserver does not).

Reply:
There are two sides: server and client. The client would be using the correct version of the API, but the server may have a backwards incompatible change where it does not serve the correct payload to the client. In turn, the client may have a backwards incompatible change, and not know what to do with a particular payload. So there needs to be two checks: one in the MCO **and** one within the API. Each component has a responsibility to check the skew, since either side could break.

Reply:
I strongly disagree with the server having to know its clients. This is not how it works in any of the API life cycle policies we apply in Kubernetes and OpenShift. The server has O(100) clients in an OpenShift cluster. It cannot follow individual skew policies, and I don't see why we should introduce another API philosophy here.

> a backwards incompatible change

This is disallowed for the server and hence never happens (if we are doing our job well, and nearly always did so in the past).

Reply:
We're not asking the server to know its clients though. We're asking its Operator to know them.

Reply:
To recap the discussion we had on this for those that were not able to attend:

- The API server ships with a set of built-in types (pod and node).
- The upstream project via API review/human or project operator oversight ensures that skew between API server and kubelet is n-2.
- The API server operator is being asked to report Upgradeable=False to capture that API review/human/operational oversight because it is the component that must not upgrade.
- The Machine Config Operator is the component that must upgrade, so reporting Upgradeable=False is confusing.
- The Machine Config Operator _should_ report a message in its reporting that admins should perform an action.

I agree that the API server (and its operators) should not have to understand all API types it serves and all its clients.  I think its reasonable for API types that are built-in, whose upstream set of human operators/api-reviewers/etc. are trying to ensure that skew in either direction (api-server or kubelet) is preserved that both operators in our context should report a desired action (apiserver operator should say i cannot upgrade, machine config operator should promote an upgrade). 

For types that are not built-in (CRDs, etc.), it is extremely reasonable to say that the future philosophy is not for the API server to block its own upgrade on those types because the upgrade of the API server itself is not what will evolve those CRD definitions, unlike the built-in types.